Sender.java

```diff
@@ -0,0 +1,206 @@
+                ObjectInputStream readFromNet = new ObjectInputStream(joinConnection.getInputStream());
+
+                  //send join request
+                  writeToNet.writeObject(new JoinMessage(ChatNode.self));
+
+                  //done
+                  joinConnection.close();
+                }
+                  catch (IOException ex)
+                  {
+                    System.out.println(ex.toString());
+                  }
+                  //send JOINED message to SUCCESSORs PREDECESSOR
+                  Socket joinedConnection = null;
+                  try
+                  {
+
+                    for( int index = 0; index < ChatNode.participantList.size(); index++ )
+                    {
+                      joinedConnection = new Socket(ChatNode.self.ip, ChatNode.self.port);
+                      ObjectOutputStream writeToNet = new ObjectOutputStream(joinedConnection.getOutputStream());
+                      writeToNet.writeObject(new JoinMessage(ChatNode.self));
+                      joinedConnection.close();
+                    }
+
+
+                  } catch (IOException ex) {
+                    System.out.println(ex.toString());
+                    continue;
+                  }
+                  hasJoined = true;
+
+              }
+
+              // if starts with LEAVE
+              else if(inputLine.startsWith("LEAVE") || inputLine.startsWith("SHUTDOWN"))
+              {
+                if (hasJoined == false)
+                {
+                  System.err.println("You have not joined a chat yet ...");
+                  continue;
+                }
+
+                // leaving chat
+                LeaveMessage leaveMessage = new LeaveMessage(ChatNode.self);
+                Socket leaveConnection = null;
+
+                try {
+                  for( int index = 0; index < ChatNode.participantList.size(); index++ )
+                  {
+                    leaveConnection = new Socket(ChatNode.self.ip, ChatNode.self.port);
+                    ObjectOutputStream writeToNet = new ObjectOutputStream(leaveConnection.getOutputStream());
+                    writeToNet.writeObject(leaveMessage);
+                    leaveConnection.close();
+                  }
+
+                } catch (IOException ex) {
+                  System.out.println(ex.toString());
+                  continue;
+                }
+
+                if (inputLine.startsWith("LEAVE"))
+                {
+                  hasJoined = false;
+                  System.out.println("Left chat");
+                }
+                else
+                {
+                  // shut down
+                  System.out.println("Shutting down\n");
+                  System.exit(0);
+                }
+              }
+
+              // otherwise, send message
+
+              else
+              {
+                if(hasJoined = false)
```

Comment:
I detect that this code is problematic. According to the [Correctness (CORRECTNESS)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#correctness-correctness), [QBA: Method assigns boolean literal in boolean expression (QBA_QUESTIONABLE_BOOLEAN_ASSIGNMENT)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#qba-method-assigns-boolean-literal-in-boolean-expression-qba-questionable-boolean-assignment).
This method assigns a literal boolean value (true or false) to a boolean variable inside an if or while expression. Most probably this was supposed to be a boolean comparison using ==, not an assignment using =.